### PR TITLE
HPCC-26560 Fix kj/stats. race

### PR DESCRIPTION
**Pull request by jakesmith**
Signed-off-by: Jake Smith <jake.smith@lexisnexisrisk.com>

## Type of change:
- [x] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [ ] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [ ] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

### CODE REVIEW

**Review by shamser — APPROVED** [review]

**Comment by jakesmith** [comment]
There was a failure in keyed_denormalize, but I'm not convinced related to this PR.
I can't reproduce the issue or see why it would happen.
Part of the testnig, I've repeatedly run this test in GH, e.g. : https://github.com/hpcc-systems/HPCC-Platform/runs/3685464084
but can't reproduce the error that way either.

The error suggests that createSegmentMonitors is being called on same IKeyManager without a corresponding releaseSegmentMonitors.
Given the variant of keyed_denormalize that was running when it failed:
```
//version multiPart=true,useTranslation=true,forceRemoteKeyedLookup=true,forceRemoteKeyedFetch=true
```

it has to have been the TLK that was involved, and therefore the createSegmentMonitors in queueLookup(const void *lhsRow) @ line 2485.
This is processing LHS rows, finding the correct part/handler, and is single threaded (called via readAhead())
From the error, and the variety of the test

**Comment by jakesmith** [comment]
> Given the variant of keyed_denormalize that was running when it failed:
> 
> ```
> //version multiPart=true,useTranslation=true,forceRemoteKeyedLookup=true,forceRemoteKeyedFetch=true
> ```
> 
> it has to have been the TLK that was involved, and therefore the createSegmentMonitors in queueLookup(const void *lhsRow) @ line 2485.
> This is processing LHS rows, finding the correct part/handler, and is single threaded (called via readAhead())

Thanks to @ghalliday - I realized I was looking in the wrong area. The exception happened on the remote service side.
It is still probably a rare ace condition which I'll keep investigating, but unrelated to this PR.

File: thorlcr/activities/keyedjoin/thkeyedjoinslave.cpp [hunk 1]
```diff
@@ -3196,8 +3214,8 @@ class CKeyedJoinSlave : public CSlaveActivity, implements IJoinProcessor, implem
         endOfInput = false;
         lookupThreadLimiter.reset();
         fetchThreadLimiter.reset();
-        keyLookupHandlers.init();
-        fetchLookupHandlers.init();
+        keyLookupHandlers.reset();
+        fetchLookupHandlers.reset();
```
Comment (ghalliday) [inline thorlcr/activities/keyedjoin/thkeyedjoinslave.cpp:3218]: trivial: extra whitespace.

**Review by ghalliday — APPROVED** [review]
@jakesmith I think this looks ok, although I doubt I would spot any fundamental flaws.  Please squash and cleanup the trailing whitespace.
Are you going to release a fix for not calling releaseSegments in the remote code as a separate jira, or decided it didn't need to change?

**Comment by jakesmith** [comment]
@ghalliday - removed whitespace and squashed

**Comment by jakesmith** [comment]
> Are you going to release a fix for not calling releaseSegments in the remote code as a separate jira, or decided it didn't need to change?

I opened a separate JIRA: https://track.hpccsystems.com/browse/HPCC-26593 (linked to this one).
It probably is worth it, but caching will usually mean it will have little effect.

**Pull request merged**